File: leetcode/Hard/0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
  vector<int> maxSlidingWindow(vector<int> &nums, int k) {
    deque<int> q;
    vector<int> ans;

    for (int i = 0; i < nums.size(); ++i) {
      cleanDeque(q, nums, i, k);
      q.push_back(i);
      if (i >= k - 1) {
        ans.push_back(nums[q.front()]);
      }
    }

    return ans;
  }

  void cleanDeque(deque<int> &q, vector<int> &nums, int i, int k) {
    if (!q.empty() && q.front() == i - k) {
      q.pop_front();
    }
    while (!q.empty() && nums[i] >= nums[q.back()]) {
      q.pop_back();
    }
  }
};

File: leetcode/Hard/0239-sliding-window-maximum/0239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque()
        ans = []

        for i in range(len(nums)):
            self.clean_deque(q, nums, i, k)
            q.append(i)
            if i >= k - 1:
                ans.append(nums[q[0]])

        return ans

    def clean_deque(self, q, nums, i, k):
        if q and q[0] == i - k:
            q.popleft()
        while q and nums[i] >= nums[q[-1]]:
            q.pop()
